deepchem/models/jax_models/jax_model.py

```diff
@@ -0,0 +1,373 @@
+      For the time being, it is optax object
+    log_frequency: int, optional (default 100)
+      The frequency at which to log data. Data is logged using
+      `logging` by default.
+
+    Miscellanous Parameters Yet To Add
+    ----------------------------------
+    model_dir: str, optional (default None)
+      Will be added along with the save & load method
+    tensorboard: bool, optional (default False)
+      whether to log progress to TensorBoard during training
+    wandb: bool, optional (default False)
+      whether to log progress to Weights & Biases during training
+
+    Work in Progress
+    ----------------
+    [1] Integerate the optax losses, optimizers, schedulers with Deepchem
+    [2] Support for saving & loading the model.
+    """
+    super(JaxModel, self).__init__(model=model, **kwargs)
+    warnings.warn(
+        'JaxModel is still under work and hence a lot of the method might not be implemented'
+    )
+    self._loss_fn = loss  # lambda pred, tar: jnp.mean(optax.l2_loss(pred, tar))
+    self.batch_size = batch_size
+    self.learning_rate = learning_rate
+    self.optimizer = optimizer
+    self.model = model
+    self.params = params
+    self._built = False
+    self.log_frequency = log_frequency
+
+    if output_types is None:
+      self._prediction_outputs = None
+      self._loss_outputs = None
+      self._variance_outputs = None
+      self._other_outputs = None
+    else:
+      self._prediction_outputs = []
+      self._loss_outputs = []
+      self._variance_outputs = []
+      self._other_outputs = []
+      for i, type in enumerate(output_types):
+        if type == 'prediction':
+          self._prediction_outputs.append(i)
+        elif type == 'loss':
+          self._loss_outputs.append(i)
+        elif type == 'variance':
+          self._variance_outputs.append(i)
+        else:
+          self._other_outputs.append(i)
+      if len(self._loss_outputs) == 0:
+        self._loss_outputs = self._prediction_outputs
+
+  def _ensure_built(self):
+    """The first time this is called, create internal data structures.
+
+    Work in Progress
+    ----------------
+    [1] Integerate the optax losses, optimizers, schedulers with Deepchem
+
+    """
+    if self._built:
+      return
+
+    self._built = True
+    self._global_step = 0
+    self.opt_state = self.optimizer.init(self.params)
+
+  def fit(self,
+          dataset: Dataset,
+          nb_epochs: int = 10,
+          deterministic: bool = False,
+          loss: Union[Loss, LossFn] = None,
+          callbacks: Union[Callable, List[Callable]] = [],
+          all_losses: Optional[List[float]] = None) -> float:
+    """Train this model on a dataset.
+
+    Parameters
+    ----------
+    dataset: Dataset
+      the Dataset to train on
+    nb_epoch: int
+      the number of epochs to train for
+    deterministic: bool
+      if True, the samples are processed in order.  If False, a different random
+      order is used for each epoch.
```

Comment:
Is this fixing the seed so that the batches are exactly the same between training runs?

Reply:
This option isn't to get the same behavior between runs.  To do that, you would fix the random number seeds.  I'm honestly not sure when someone would want to set this option to True.  Using a different order for each epoch usually gives better training.  Maybe there's some situation I haven't thought of where someone would want the same order for every epoch.

Reply:
1) Oh my bad, thanks for letting me know. I had gotten confused with `seed` value being passed here https://github.com/deepchem/deepchem/blob/66402d1cf34d51c13ed491f1f14eaffca133dfab/deepchem/splits/splitters.py#L369

2) Yes, I agree that having different orders of batches during training will definitely give better training. The precise results of a model can only be reproduced if the random training order of batches and initial model weights are the same. With my limited exposure to `Jax`, it has an amazing ability to reproduce results because of the way in which it generates its random numbers. We were discussing this, merely to check this ability to reproduce very well in Jax. In my runs, I wasn't able to reproduce my results on different instances of collab sessions.

Based on your suggestions, I think setting `random.random(seed)` might fix the problem.

Reply:
And also `np.random.seed()`.  Actually, I don't think we ever use `random` anywhere.  But we do use `np.random`.